tasks/shell.sh

```diff
@@ -0,0 +1,555 @@
+
+# CODEREVIEW: What are we doing with this?
+if [ -f /opt/puppetlabs/puppet/VERSION ]; then
+  installed_version=`cat /opt/puppetlabs/puppet/VERSION`
+else
+  installed_version=uninstalled
+fi
+
+# Retrieve Platform and Platform Version
+if test -f "/etc/lsb-release" && grep -q DISTRIB_ID /etc/lsb-release; then
+  platform=`grep DISTRIB_ID /etc/lsb-release | cut -d "=" -f 2 | tr '[A-Z]' '[a-z]'`
+  platform_version=`grep DISTRIB_RELEASE /etc/lsb-release | cut -d "=" -f 2`
+elif test -f "/etc/debian_version"; then
+  platform="debian"
+  platform_version=`cat /etc/debian_version`
+elif test -f "/etc/redhat-release"; then
+  platform=`sed 's/^\(.\+\) release.*/\1/' /etc/redhat-release | tr '[A-Z]' '[a-z]'`
+  platform_version=`sed 's/^.\+ release \([.0-9]\+\).*/\1/' /etc/redhat-release`
+
+  #If /etc/redhat-release exists, we act like RHEL by default. Except for fedora
+  if test "$platform" = "fedora"; then
+    platform="fedora"
+  else
+    platform="el"
+  fi
+elif test -f "/etc/system-release"; then
+  platform=`sed 's/^\(.\+\) release.\+/\1/' /etc/system-release | tr '[A-Z]' '[a-z]'`
+  platform_version=`sed 's/^.\+ release \([.0-9]\+\).*/\1/' /etc/system-release | tr '[A-Z]' '[a-z]'`
+  # amazon is built off of fedora, so act like RHEL
+  if test "$platform" = "amazon linux ami"; then
+    platform="el"
+    platform_version="6.0"
+  fi
+# Apple OS X
+elif test -f "/usr/bin/sw_vers"; then
+  platform="mac_os_x"
+  # Matching the tab-space with sed is error-prone
+  platform_version=`sw_vers | awk '/^ProductVersion:/ { print $2 }'`
+
+  major_version=`echo $platform_version | cut -d. -f1,2`
+  case $major_version in
+    "10.6") platform_version="10.6" ;;
+    "10.7"|"10.8"|"10.9") platform_version="10.7" ;;
+    *) echo "No builds for platform: $major_version"
+       report_bug
+       exit 1
+       ;;
+  esac
+
+  # x86_64 Apple hardware often runs 32-bit kernels (see OHAI-63)
+  x86_64=`sysctl -n hw.optional.x86_64`
+  if test $x86_64 -eq 1; then
+    machine="x86_64"
+  fi
+elif test -f "/etc/release"; then
+  platform="solaris2"
+  machine=`/usr/bin/uname -p`
+  platform_version=`/usr/bin/uname -r`
+elif test -f "/etc/SuSE-release"; then
+  if grep -q 'Enterprise' /etc/SuSE-release;
+  then
+      platform="sles"
+      platform_version=`awk '/^VERSION/ {V = $3}; /^PATCHLEVEL/ {P = $3}; END {print V "." P}' /etc/SuSE-release`
+  else
+      platform="suse"
+      platform_version=`awk '/^VERSION =/ { print $3 }' /etc/SuSE-release`
+  fi
+elif test -f "/etc/arch-release"; then
+  platform="archlinux"
+  platform_version=`/usr/bin/uname -r`
+elif test "x$os" = "xFreeBSD"; then
+  platform="freebsd"
+  platform_version=`uname -r | sed 's/-.*//'`
+elif test "x$os" = "xAIX"; then
+  platform="aix"
+  platform_version=`uname -v`
+  machine="ppc"
+fi
+
+if test "x$platform" = "x"; then
+  critical "Unable to determine platform version!"
+  report_bug
+  exit 1
+fi
+
+# TODO: we shold either get a specific version or latest right?
```

Comment:
Can I remove all of this logic and force the user specify a specific puppet agent package version?

Reply:
We specifically wanted that use case because it seemed like the most common request would be: I want a Puppet 5 agent package, give me the latest. The logic was specifically
- if I don't specify a version and nothing's installed, install latest
- if I don't specify a version and something's installed, leave it
- if I specify 'latest', install or update to the latest version
- if I specify a version, install or update to that version